scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -436,112 +445,232 @@
+                                                                (.getAssignedPorts task))))
+                                               tasks)
+                              slave-id (:slave-id (first offers))]]
+                  (mesos/launch-tasks
+                    driver
+                    (mapv :id offers)
+                    (mapv #(-> %
+                               (assoc :slave-id slave-id)
+                               (dissoc :num-ports))
+                          task-infos))
+                  (doseq [^TaskAssignmentResult task tasks]
+                    (.. fenzo
+                        (getTaskAssigner)
+                        (call (.getRequest task) (get-in (first leases) [:offer :hostname])))))
+                (:matched-head? processed-matches)))))
       (catch Throwable t
-        (decline-offers driver (:ids offer))
         (meters/mark! handle-resource-offer!-errors)
-        (log/error t "Error in match:" (ex-data t))))))
+        (async/go
+          (async/>! offers-chan @offer-stash))
+        (log/error t "Error in match:" (ex-data t)))))))
+
+(defn view-incubating-offers
+  [^TaskScheduler fenzo]
+  (let [pending-offers (for [^com.netflix.fenzo.VirtualMachineCurrentState state (.getVmCurrentStates fenzo)
+                             :let [lease (.getCurrAvailableResources state)]
+                             :when lease]
+                         {:hostname (.hostname lease)
+                          :slave-id (.getVMID lease)
+                          :resources {:cpus (.cpuCores lease)
+                                      :mem (.memoryMB lease)}})]
+    (log/debug "We have" (count pending-offers) "pending offers")
+    pending-offers))
+
+(defn make-offer-handler
+  [conn driver-atom fenzo fid-atom pending-jobs-atom]
+  (let [offers-chan (async/chan (async/buffer 5))
+        resources-atom (atom (view-incubating-offers fenzo))
+        timer-chan (chime-ch (periodic/periodic-seq (time/now) (time/seconds 1))
+                             {:ch (async/chan (async/sliding-buffer 1))})
+        max-considerable 1000]
+    (async/thread
+      (loop [num-considerable max-considerable]
+        ;;TODO make this cancelable (if we want to be able to restart the server w/o restarting the JVM)
+        (recur
+          (try
+            (let [offers (async/alt!!
+                           offers-chan ([offers] offers)
+                           timer-chan ([_] [])
+                           :priority true)
+                  matched-head? (handle-resource-offers! conn @driver-atom fenzo @fid-atom pending-jobs-atom num-considerable offers-chan offers)]
+              (when (seq offers)
+                (reset! resources-atom (view-incubating-offers fenzo)))
+              (if matched-head?
+                max-considerable
+                (max 1 (long (* 0.95 num-considerable))))) ;; With max=1000 and 1 iter/sec, this will take 88 seconds to reach 1
```

Comment:
@wyegelwel  I believe it would vary widely depending on the cluster configuration and workload.   Perhaps we should record a metric about the number of iterations required before one is accepted?